library/kani_macros/src/sysroot/contracts.rs

```diff
@@ -794,13 +861,41 @@ fn requires_ensures_main(attr: TokenStream, item: TokenStream, is_requires: bool
             let ItemFn { attrs, vis, sig, block } = &item_fn;
             handler.output.extend(quote!(
                 #(#attrs)*
-                #[kanitool::checked_with = #check_fn_name_str]
+                #[kanitool::checked_with = #recursion_wrapper_name_str]
                 #[kanitool::replaced_with = #replace_fn_name_str]
                 #vis #sig {
                     #block
                 }
             ));
 
+            let mut wrapper_sig = sig.clone();
+            attach_require_kani_any(&mut wrapper_sig);
```

Comment:
This is mentioned in the callouts, but I'm not clear on how this constraints things beyond the already existing call to `attach_require_kani_any` in `emit_replace_function`.

Reply:
Well, the wrapper calls both check and replace, and since replace requires in `impl kani::Arbitary` so must the wrapper

Reply:
Can you give an example of a function that would be rejected, that was accepted before this PR?

Reply:
Oh yes, so if you had `Option::unwrap` for instance

```rs
impl<T> Option<T> {
  #[kani::ensures(...doesnt matter)]
  fn unwrap(self) -> T {
    if let Some(t) = self {
      t
    } else { panic!() }
  }
}
```

Then before the signature of the check function would have been `for<T> fn(Option<T>) -> T` and of replace `for<T: Arbitrary> fn(Option<T>) -> T`.
In other words you could have done `proof_for_contract` for your custom `struct MyThing { .. }` and if you wanted to use `stub_verified` you'd have to implement `Arbitrary for MyThing` (makes sense how else to would you do it).

But now "check" (or in this case "wrapper" which is now the new entry to the checking logic) calls "replace" and so the signature is `for<T: Arbitrary> fn(Option<T>) -> T` even for the check so if you want to do `proof_for_contract` you need `Arbitrary for MyThing`.

Reply:
I see. In that case, I think it makes sense to require `Arbitrary` anyway, since there one is likely to want to use `stub_verified`, and not just `proof_for_contract`.

Reply:
Probably true but I've opened #2823 to maybe relax this restriction.